src/EditorFeatures/Core/LanguageServer/Handlers/Completion/CompletionHandler.cs

```diff
@@ -169,101 +185,85 @@ bool IsValidTriggerCharacterForDocument(Document document, char triggerCharacter
                 return true;
             }
 
-            static async Task<LSP.CompletionItem> CreateLSPCompletionItemAsync(
-                LSP.CompletionParams request,
-                Document document,
-                CompletionItem item,
-                CompletionResolveData? completionResolveData,
-                bool supportsVSExtensions,
-                Dictionary<ImmutableArray<CharacterSetModificationRule>, string[]> commitCharacterRulesCache,
-                CompletionService completionService,
-                bool returnTextEdits,
-                bool snippetsSupported,
-                StringBuilder stringBuilder,
-                SourceText? documentText,
-                TextSpan? defaultSpan,
-                LSP.Range? defaultRange,
-                CancellationToken cancellationToken)
+            static void AddRoslynItemProperties(LSP.CompletionItem lspItem, CompletionItem roslynItem, StringBuilder stringBuilder, CompletionResolveData? completionResolveData)
             {
                 // Generate display text
-                stringBuilder.Append(item.DisplayTextPrefix);
-                stringBuilder.Append(item.DisplayText);
-                stringBuilder.Append(item.DisplayTextSuffix);
+                stringBuilder.Append(roslynItem.DisplayTextPrefix);
+                stringBuilder.Append(roslynItem.DisplayText);
+                stringBuilder.Append(roslynItem.DisplayTextSuffix);
                 var completeDisplayText = stringBuilder.ToString();
                 stringBuilder.Clear();
 
-                var completionItem = supportsVSExtensions ? new LSP.VSInternalCompletionItem() : new LSP.CompletionItem();
-                completionItem.Label = completeDisplayText;
-                completionItem.SortText = item.SortText;
-                completionItem.FilterText = item.FilterText;
-                completionItem.Kind = GetCompletionKind(item.Tags);
-                completionItem.Data = completionResolveData;
-                completionItem.Preselect = ShouldItemBePreselected(item);
+                lspItem.Label = completeDisplayText;
+                lspItem.SortText = roslynItem.SortText;
+                lspItem.FilterText = roslynItem.FilterText;
+                lspItem.Kind = GetCompletionKind(roslynItem.Tags);
+                lspItem.Data = completionResolveData;
+                lspItem.Preselect = ShouldItemBePreselected(roslynItem);
 
+                if (lspItem is LSP.VSInternalCompletionItem vsCompletionItem)
+                {
+                    vsCompletionItem.Icon = new ImageElement(roslynItem.Tags.GetFirstGlyph().GetImageId());
+                }
+            }
+
+            static async Task AddTextEditsAsync(
```

Comment:
Given this doesn't always add a text edit and occasionally depends on Resolve to do the needful I wonder if there's a better name we could use to represent that